compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -9205,11 +9303,22 @@ TR::Node *bdivSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
    {
    s->simplifyChildren(node, block);
 
+   // Handshake with divchkSimplifier:  If simplifying the child of a DIVCHK
+   // results in a node that still needs to have a DIVCHK applied, place that
+   // node in _nodeToDivchk.  If the simplification leaves no node that needs
+   // to have a DIVCHK applied, set _nodeToDivchk to NULL.
+   s->_nodeToDivchk = node;
```

Comment:
Fixed this in `BINARY_IDENTITY_OP` in commit 37c05e256709f96c1913b8fcab1b45b3e3b2dca2.  It looks like that macro is getting a bit unwieldy with this change.  I'm wondering whether it should become a function - the fact that the macro returns from the "caller" was already a bit yucky.  Thoughts?

Reply:
I agree it would be better for `BINARY_IDENTITY_OP()` to become a function. It's not only the hidden control flow - it also implicitly refers to local variables by name (`firstChild`, `secondChild`)

Whether we change it to a function or leave it as-is, it might be better to move the test for a constant `firstChild` to a nested `if` that guards just the constant folding, and the `BINARY_IDENTITY_OP()` part could move into the outer `if`, e.g.
```
if (second is const && second nonzero && performTransformation(...))
   {
   s->_nodeToDivchk = NULL;
   if (first is const)
      {
      fold...Constant(...);
      return node;
      }

   BINARY_IDENTITY_OP(..., 1)
   }
```
That way `BINARY_IDENTITY_OP()` wouldn't need to concern itself with `_nodeToDivchk`, and the outer `performTransformation()` would be more analogous to the ones in the 32- and 64-bit div/rem simplifiers

What do you think?

Reply:
Moved uses of `BINARY_IDENTITY_OP` for division and remainder simplifiers divisor is known to be a non-zero constant in commit 04b2b82964f525e4f3c4d7c88a7fb0ec9d1a1b21